File: 1035-cousins-in-binary-tree/cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void dfs(TreeNode root,int arr[],int parent,int depth,int x,int y){
        if(root==null) return;
        if(root.val==x){
            arr[0] = depth;
            arr[1] = parent;
        }
        if(root.val==y){
            arr[2] = depth;
            arr[3] = parent;
        }
        dfs(root.left,arr,root.val,depth+1,x,y);
        dfs(root.right,arr,root.val,depth+1,x,y);
    }
    public boolean isCousins(TreeNode root, int x, int y) {
        int arr[] = new int[4];
        dfs(root,arr,root.val,1,x,y);
        return arr[0]==arr[2] && arr[1]!=arr[3];
    }
}

File: 1035-cousins-in-binary-tree/cousins-in-binary-tree.cpp
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        if (!root) return false;
        
        queue<pair<TreeNode*, pair<int, int>>> q;
        q.push({root, {-1, 0}});
        
        int x_parent = -1, y_parent = -1, x_depth = -1, y_depth = -1;
        
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                auto [node, info] = q.front(); q.pop();
                int parent = info.first, depth = info.second;
                
                if (node->val == x) {
                    x_parent = parent;
                    x_depth = depth;
                }
                if (node->val == y) {
                    y_parent = parent;
                    y_depth = depth;
                }
                
                if (node->left) q.push({node->left, {node->val, depth + 1}});
                if (node->right) q.push({node->right, {node->val, depth + 1}});
            }
            
            if (x_parent != -1 && y_parent != -1) break;
        }
        
        return (x_depth == y_depth) && (x_parent != y_parent);
    }
};